### ISSUE TITLE
Blitz-Out Inconsistency (2.5.22294)

### ISSUE THREAD

**Issue by Cernelius** [triplea-game/triplea#10643]
Situation:
A unit which is able to blitz moves out of an (enemy-owned) territory into an other land zone, and the territory has no enemy units in it (It does not matter what units are in the other land zone.).

Correct behaviour:
The territory (of which the unit is moving out) ought (always) to be conquered as if blitzed.

Current behaviour:
The territory is conquered if the other land zone is an other enemy-owned territory but is not conquered if the other land zone is friendly.

Savegame:
[2022-6-12-World-War-II-Revised(blitz out).zip](https://github.com/triplea-game/triplea/files/8886331/2022-6-12-World-War-II-Revised.blitz.out.zip)

Additional information:
You can test all the following behaviours by moving one of the units which are in West Russia:

1. Move the armour to Karelia S.S.R.: West Russia is not conquered.
2. Move the armour to Belorussia: West Russia is conquered.
3. Move the infantry to Karelia S.S.R.: West Russia is not conquered.
4. Move the infantry to Belorussia: impossible.

The behaviour described at the first point is wrong (in that West Russia should be conquered), whereas all other behaviours are correct, so the first behaviour should be changed as to have West Russia being conquered upon exiting but not changing the other behaviours (In particular, West Russia should not be conquered when the infantry moves out because the infantry cannot blitz.).

Notice that this has nothing to do with the armour being able to move two spaces: what matters is the `canBlitz` option being set true, not the movement: the armour would behave here just like the infantry if the armour would have been able to move 2 spaces but unable to blitz.

Please fix.

**Comment by Cernelius** [comment]
@asvitkine May you fix this? I think it is strongly related to what you are doing at
https://github.com/triplea-game/triplea/pull/10617

**Comment by panther2** [comment]
@Cernelius 

Please help me understand this scenario. How can Russia start a Combat Move in an enemy territory?

**Comment by Cernelius** [comment]
> @Cernelius
> 
> Please help me understand this scenario. How can Russia start a Combat Move in an enemy territory in Revised? Isn't the situation rather caused by editing an impossible scenario to a map not supporting something you intend?

Yes, this would never actually happen in this game or any other basic ones. However, it rests on basic rules.
Since entering an empty enemy territory and going back to the territory you came from is a blitzing move (even though ending into a friendly territory) and causes the territory to be conquered just like if we move through it into an other enemy territory, then the outcome at the first point should be the same as the outcome at the second point (West Russia being captured) because it should not matter whether the final territory is friendly or hostile. I think it can be assumed that, if a unit can blitz through a territory, the same unit should be able to blitz out of a territory, which is how the program already behaves in the second point (If you select the armour in West Russia and move it to Belorussia, you will see that both territories are immediately conquered, which is an agreeable outcome beside the inconsequential and notorious fact that Belorussia should be actually conquered during Conduct Combat if the armour does not move any further.) but, inconsistently, not in the first one.
An alternative solution would be having West Russia non-conquered both at the first and at the second point, but it would make no sense that blitzing a territory you are already inside is harder than blitzing an adjacent one (as, this way, the only way to blitz West Russia would be having an armour that can move 3 or more spaces so that you first move out of West Russia and then move into it again and then out of it again, making a total movement like West Russia to Karelia S.S.R., then Karelia S.S.R. to West Russia, then West Russia to Karelia S.S.R. (This also does not matter for the basic games because it would require land units able to move 3 spaces or more, which are, however, quite common amongst TripleA games.). The absurdity would be that, in the given save-game, an armour starting in Karelia S.S.R. would require only 2 movements to take both West Russia and Belorussia and end its movement in Belorussia (by moving Karelia S.S.R. to West Russia to Belorussia), whereas an armour starting in West Russia would require at least 3 movements to take both West Russia and Belorussia and end its movement in Belorussia, and it would make no sense that West Russia is harder to blitz if you start your turn inside it than if you start your turn next to it.

Do you agree with my (let's say) "extrapolation" of the rules to this case which can never actually happen in the basic games (but certainly can in other TripleA games)? The main thing which I believe we can certainly agree is wrong is the inconsistency between the first and the second points: West Russia should either be conquered in both cases or in neither of them. This said, we may then agree that the preferable solution is West Russia being conquered in both cases, which is my point. Otherwise, a property could be added by which West Russia would be conquered in both cases if false and in neither of them if true. In any case, the current behaviour is inconsistent and should be changed.

**Comment by panther2** [comment]
@Cernelius 

Abstracting from the basic games I think I understand your point now.

So let's assume a game, which rules allow units starting in an enemy territory.
(Maybe there has been a battle in West Russia one round before and on Germany's turn 
Germany retreated its units to Ukraine, as the rules of that game allow that.)

In this case Russia starts its turn in West Russia, that at this point has not changed ownership 
to Russia.

Now the question is: At which point does a territory change ownership. Abstracting a bit from Revised and enhancing the question to modern core rules (v3-v6 incl. 1940) the rules say that ownership ususally changes at the end of the Conduct Combat phase (during the Conclude Combat step). Only in the special case of a Blitz through an enemy territory the territory blitzed through changes ownership immediately before the blitzing unit continues its move to a second (friendly or hostile) territory.

Now in your scenario we must first be clear about two different points:

- when does an Enemy Territory change ownership in the situation that leads to your starting position of your savegame?
- what is related to blitz and what is not?

I leave the first point open for now. As this must be defined by the mapmaker with his intended rules. There are of course several option.

Concerning the second point I would say that - if the rules allow this starting position - the Russian infantry should be allowed to move its one point movement to any enemy territory. It should not be allowed to move into friendly territories, as we are in Combat Move Phase. The Infantry's move is totally unrelated to blitzing.
The Russian Tank however is a unit allowed to blitz, so the Tank should be able to enter an enemy territory as his first move (here Belorussia), taking this territory immediately when continuing his move to a second territory (hostile or friendly). This - and only this - is bltzing.

So it seems to me that the question of West Russia changing owership is rather unrelated to the question of "blitzing".
To me it seems that it is necessary to clearly define changing ownership of territories in special senarios like this one before.

It might for example be defined that
- ownership changes when German Units move out or at the end of Germany's turn, leaving the territory to Russia
- ownership changes at the beginning of Russia's turn or Combat Move
- ownership changes during Russia's turn after one round of combat against no one 
- whatever ...

**Comment by panther2** [comment]
@asvitkine The above zip file contains the*tsvg file. Any zip-software should be able to extract it:
See:
![zip](https://user-images.githubusercontent.com/20812775/173382705-fe8483e5-7719-4c65-a0e6-f807922190ea.png)

**Comment by Cernelius** [comment]
> I agree with @panther2 here, I think the state in that save game, which I assume was created with Edit Mode, can't really happen regularly, so now we're talking about custom map territories.

This (The fact that I obtained it with edit mode instead of by playing a game regularly.) doesn't matter: it is something that can happen in a number of TripleA games. I could have obtained the functionally similar situation by starting Napoleonic Empires and playing it (without any edits) till getting to that, and the only difference would have been spending more time to do so.

> So then, I'd treat it more as a feature request to improve the behavior for maps that may want that, as opposed to a bug per se since it won't happen regularly.

The problem is (as I said) that the current behaviour is inconsistent. Can you explain me why I'm conquering West Russia when I move from West Russia to Belorussia but not when I move from West Russia to Karelia S.S.R.? I don't think there is a sensible explanation for that and it is also inconsistent with the fact that regular blitzing happens all the same in case I'm moving friendly - hostile - hostile and in case I'm moving friendly - hostile - friendly (for example, if I have an armour starting in Karelia S.S.R. (in the save-game), I blitz West Russia both in case I move Karelia S.S.R. - West Russia - Belorussia and in case I move Karelia S.S.R. - West Russia - Karelia S.S.R..
The inconsistency is the problem.
How to fix the inconsistency is not obvious, however, since this never happens in the basic games, but this does not mean there is not a problem: just that there is not a given solution to it.

> Having said that, it looks like my blitz PR may change things here a bit, for example it does let the infantry move out of that territory into either of the adjacent enemy territories (since it's not related to blitzing). However, perhaps it shouldn't be doing that - since we want to default "can attack from contested territory" to false, so I think if we're treating this territory as "contested", it should prevent such movement (unless the property is set).

I don't think you should be changing any such behaviours unless you are doing it deliberately and documenting it in PoS2. I think most such behaviours have no documentation nor rules-books behind, so can be changed, but only if this springs from a general plan on how to define them (hopefully better than they are). The next step would be documenting everything in PoS2, so that people may know that how it works is how it is intended to work and won't have to test themselves every case to know how it works.

This said, I think point 4 ("Move the infantry to Belorussia: impossible.") is a preferable behaviour. I think an infantry starting in a hostile territory must conquer that territory before conquering anything else but it can conquer nothing else than one territory because it cannot blitz, so the only way for moving the infantry would be point 3 ("Move the infantry to Karelia S.S.R.: West Russia is not conquered."), which I consider a correct behaviour too. If you intend to change either or both of such behaviours, I much suggest you add one or two properties to allow map-makers to go either way, preferably keeping the current behaviour as default.
As I said, I tend to think they should better remain working as they are, but it has to be open to discussion.

**Comment by asvitkine** [comment]
@Cernelius looks like we replied almost at the same time. Can you check my message above and lmk if you agree with my thoughts and plan here?

**Comment by Cernelius** [comment]
> @Cernelius looks like we replied almost at the same time. Can you check my message above and lmk if you agree with my thoughts and plan here?

Of course. I've also to reply to @panther2, since I disagree with something he said.

**Comment by panther2** [comment]
@Cernelius 
Can you please do me a favour and explain to me how a similar situation can occur in Napoleonic Empires?
I have played this map a lot years ago, and fail to remember how this could happen.

**Comment by Cernelius** [comment]
> > @Cernelius Can you please do me a favour and explain to me how a similar situation can occur in Napoleonic Empires? I have played this map a lot years ago, and fail to remember how this could happen.
> 
> Two powers are in an Allied or Open Borders relationship. One power moves combat units into an empty territory of the other power and keep them there. The power downgrades the relationship to war.
> 
> However, this is just one way it can happen.

Sorry, I should have clarified that I was talking about the Napoleonic Empires map, not the Napoleonic Empires game: this is not possible in the Napoleonic Empires game but it is possible in all the other games of the Napoleonic Empires map. The games of that map in which this is possible are the "Napoleonic Empires: FFA 5 Player" and the "Napoleonic Empires: FFA 8 Player".

I assume you only played the Napoleonic Empires game of the Napoleonic Empires map and never the other games of the same map. The only game of that map I ever played is "Napoleonic Empires: FFA 5 Player".

**Comment by asvitkine** [comment]
@Cernelius Thanks for your detailed reply.

Let me reply to your two main points:
  1. Whether it makes sense the territory ownership change issue this in my pending PR.

I agree it should be fixed, but the code for changing territory ownership is completely in a different place than the move validation code I'm changing. So I think it should be in a separate PR. I don't disagree that fixing it makes sense.

As another point, given it's a new area of the code base for me to dig into, I think it would a worse use of my time than fixing some of the 2.6 regressions also in areas of the code unfamiliar to me, such as #10599 where the wrong number of dice are rolled or #10647 where there's error popups in 2.6.

  2. That it's better to give blitzing units the ability to leave contested territories by default.

Here, I disagree, let me explain my thinking. First, we're talking about very much an edge case. It requires a blitzing unit to start in a contested territory, which I think eliminates the vast majority of triplea maps. So the ones left are the ones with limited combat rounds (e.g. civil war), ones were politics can change ad hoc (napoleonic empires) and ones where the starting position or triggers create such a scenario. I think it's telling that we only discovered this as a result of explicit testing and never heard anything about this from players - i.e. it doesn't really happen much in practice.

So given it's such an edge case and in particular, it doesn't come up with any rule book we're trying to replicate, so it's entirely up to TripleA how this should work, I think we should strive for something that's the most maintainable and doesn't result in a lot of extra complexity just for this edge case.

In particular, my proposed "Can Attack From Contested Territories" property is very clear (given a clear definition of what contested means - but this is already well defined with the existing property "Contested Territories Produce No Income" - territory must not have enemy units).

If we instead need to make it take into account blitzing, it becomes quite a mess, because then the new property would need to instead say something like "Non-Blitz Units Can Attack From Contested Territories" - in order to preserve existing behavior. And then you'd want a "Blitz Units Can Attack From Contested Territories" if you want to separately control this for blitz units. But it's not even "contested territory" for the blitz case right now, but "blitzable territory" which has its own definition. So as you can see, this would require a lot of complexity - both to explain how it works and in terms of properties that would be needed to support customizing this. 

But then stepping back, my argument is still that this has nothing to do with blitzing, other than the current engine not making a distinction between blitzes and non-blitzes for these attacks (i.e. some existing behavior that's questionable). If blitz is defined as being able to attack multiple times in a combat move given the unit's movement, then I don't see how it's relevant to the starting territory.

The last thing I want to argue is the question of "Does it really matter?". That is, if the case doesn't really come up all that often, because existing rulesets+maps don't run into this scenario and the ones that do, still need a pretty specific case to happen (i.e. ally -> war while having units in an empty territory), then it's a bad reason to add a lot of complexity in terms of new properties and engine code. Instead, it's better to go with a simpler solution, of which "Can Attack From Contested Territories" is clearly, and the alternative is not.

**Comment by Cernelius** [comment]
> In particular, my proposed "Can Attack From Contested Territories" property is very clear (given a clear definition of what contested means - but this is already well defined with the existing property "Contested Territories Produce No Income" - territory must not have enemy units).

I feel it being far from clear and you previously named it as "Cannot attack out of a contested territory", so which one is it? Can or cannot?
For example, if I'm in a contested territory and want to move to a contested territory I own, does this property matter? Assuming the name of the property is "Can Attack From Contested Territories", does this mean I cannot move from any contested to my own territory if there are one or more enemy units in it and this property is false? I don't think this is what you want, but it is what the property is saying.

---

All considered, I think there is a very high risk that your changes will make default things worse off and possibily damaging existent non-basic games meant to work under exactly the current behaviour.
At this point, I feel like strongly suggesting you to keep the default behaviour exactly as it is now and make any changes to it via additional properties which are false as default.
Really, as I said, the only thing I think it should be changed of the current program behaviour is having blitzing unit conquering biltzable territories also when exiting them to friendly ones (which would fix the inconsistency that currently you can do this only by going to enemy territories). As far as I know, I think I would keep anything else which is default working exactly as it does now. Even this proposed change could be not made as default but by adding a property specifically for blitzing-on-exiting territories when going to friendly ones, though at this point it would make much more sense for such a property to apply to every blitzing-on-exiting case (since any such case is not part of the basic rules-sets anyway, which support only blitzing-on-moving-through).

**Comment by Cernelius** [comment]
> > > In particular, my proposed "Can Attack From Contested Territories" property is very clear (given a clear definition of what contested means - but this is already well defined with the existing property "Contested Territories Produce No Income" - territory must not have enemy units).
> > 
> > 
> > I feel it being far from clear and you previously named it as "Cannot attack out of a contested territory", so which one is it? Can or cannot?
> 
> Sorry, I mispoke in that reply. It's "Can Attack From Contested Territories" defaulting to false - I did not change the name in my PR, it was this way from when I introduced this, so that false was a reasonable default that intended to mostly match existing behavior (obviously, the blitz case we're discussing is the discrepancy).
> 
> > For example, if I'm in a contested territory and want to move to a contested territory I own, does this property matter? Assuming the name of the property is "Can Attack From Contested Territories", does this mean I cannot move from any contested to my own territory if there are one or more enemy units in it and this property is false? I don't think this is what you want, but it is what the property is saying.
> 
> Yes, it applies as well. If property says true, you can do it, otherwise you cannot. For majority of triplea maps, this again isn't relevant since it's the same story as the example we're discussing here - i.e. they won't have this state.

Such matters have always been an undocumented problems-ridden mess nobody ever wanted (in my experience) seriously to put his/her hands on, to try to make sense of, but I think you are doing it with far too heavy axe-strokes.
If you want to go the way you want to go, I much suggest you atomize everything much more specifically, creating more properties than that.
For example, under your system, it would not be possible to have a game working under the behaviour that units in an enemy-owned contested territory cannot move to an other enemy-owned territory but can move to a friendly-owned contested territory: if I understand what you are saying, either they would be able to move to both or to none (depending on the setting of your property). This is an example in which it would be much better having two properties, so that every map-maker can decide what is possible and what not, for the game. For example, a property which allows you to move from contested to enemy-owned and an other property which allows you to move from contested to friendly-but-contested.

I reiterate I suggest you fix problems (comprising inconsistencies (for inconsistencies which can go multiple ways you may want to add properties like in the case of blitzing-by-moving-out, which is currently possible only if you go to enemy territory but not if you go to friendly which makes no sense)) while otherwise keeping the program behaviour as close as possible as it currently is. Meaning leaving alone as default whatever has no problems.
If you change something which is not problematic, I highly suggest you make a property available for opting back into the current behaviour, on a per-game basis.

---

By the way, while you are at it, are you also fixing that problem that you can use paratroopers or land units on blitzable enemy-owned territories to open the way for other paratroopers? It has been discussed a number of times, but I don't recall was ever fixed. For example, here:
https://forums.triplea-game.org/topic/1561/cold-war-1965-official-thread/15?page=1

**Comment by Cernelius** [comment]
> I still don't find the argument for blitzable units being able to exit enemy territories they start in convincing, but the other thing you brought up about moving to your own territories that are contested - I went to check with 2.5, and this is indeed allowed, whereas my PR causes that to change. I do think that may have a more impactful unintended effect, so that convinced me to maybe try to preserve the old behavior as-is when the new property isn't set.

That was just an example. My point (that you are proving) is that I believe you should not go ahead changing things you don't even know how they currently work.

**Comment by Cernelius** [comment]
To be clear, I'm not against changing things, but it should be at least deliberate: you know exactly how everything works and you know exactly what you are changing and why.

**Comment by panther2** [comment]
@Cernelius 
> I assume you only played the Napoleonic Empires game of the Napoleonic Empires map and never the other games of the same map. The only game of that map I ever played is "Napoleonic Empires: FFA 5 Player".

Thank you, your assumption is correct.

**Comment by panther2** [comment]
Having read all the latest comments I don't have very much to add. Maybe let me summarize my thoughts as follows, as always based on core rulesets and the desire to argue close to them:

- As a workaround I don't have of course any objection against harmonizing movement where it makes sense.
- The problem of the current issue is unrelated to blitzing. Moving out of whatever territory is not 'blitzing'.
- The problem of this issue is the lack of a reasonable "territory changing ownership" process for special situations like this one.
- That could be resolved by the territory immediately changing ownership at the time the situation comes up.
- The core ruleset of 1914 has clear rules for territories (being empty of enemy units)  changing owenrship.
- A contested territory is a terriory with units of both factions inside. 1914 rules say "Contested: Not controlled by any power, and occupied by units of one or more powers on each side"

So I strongly vote for resolving this terriorial ownership loophole at first.

- Additionally to that I can think of movement rules (unrelated to blitzing) for units starting in contested territories.
- The core ruleset of 1914 has those rules.
